Arlo - Fixes for updated library

jwillaz opened the pull request:
## Description:
Update library requirement. Library has been updated to reduce calls to the API.  Clean up available attributes to handle NoneType cases when data is not yet retrieved from API calls.  Add signal strength sensor monitor.

**Related issue (if applicable):** fixes #9852

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** https://github.com/home-assistant/home-assistant.github.io/pull/3641

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

```yaml
# Example configuration.yaml entry
sensor:
  - platform: arlo
    monitored_conditions:
      - captured_today
      - last_capture
      - total_cameras
      - battery_level
      - signal_strength
```

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

jwillaz committed:
Reduce update calls to API. Add signal strength monitor.

homeassistant commented:
Hi @jwillaz, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#9892).

Once you do that we will be able to review and accept this pull request.

Thanks!

Danielhiversen commented:
You have some style issues: https://travis-ci.org/home-assistant/home-assistant/jobs/288363986#L538-L549

jwillaz committed:
Fix lint errors

jwillaz committed:
Fix indent

pvizeli commented on homeassistant/components/camera/arlo.py:
```diff
@@ -158,15 +161,48 @@ def disable_motion_detection(self):
         self.set_base_station_mode(ARLO_MODE_DISARMED)
 
+    @staticmethod
+    def clean_attr(attr):
```
Instead to set UNKNOWN, filter that attribute out

jwillaz replied:
@pvizeli Are attributes dynamic? If an attribute becomes available after a refresh, can it be added, or is the list of attributes set in stone once the component is initialized?  The reason for setting to UNKNOWN is that the data may not be readily available when the component is initialized. The values would get populated after an update pulls from the REST endpoints.

pvizeli commented on homeassistant/components/camera/arlo.py:
```diff
@@ -158,15 +161,48 @@ def disable_motion_detection(self):
-        self.attrs[ATTR_MIRRORED] = self._camera.get_mirror_state,
-        self.attrs[
-            ATTR_MOTION] = self._camera.get_motion_detection_sensitivity,
+        # pylint: disable=W0212
```
Not allow to disable. Thats make uggly workarounds

jwillaz replied:
@tchellomello Should we add a method to the PyArlo library classes to retrieve the session on the objects (and properties like refresh_rate and last_refresh). That would fix these lint issues (protected access).

tchellomello replied:
@jwillaz  +1 yes, let's implement. Do you want to submit a PR?

pvizeli replied:
Yes the correct work flow is: update module library -> update home-assistant

tchellomello replied:
@jwillaz  PR https://github.com/tchellomello/python-arlo/pull/60 created on `pyarlo` library.

tchellomello replied:
@jwillaz  https://pypi.python.org/pypi/pyarlo/0.0.9 has been published.

tchellomello replied:
Implemented by https://github.com/tchellomello/python-arlo/pull/61

pvizeli reviewed (changes requested):
Maybe you look to our Throttle object in utils

mikesplain commented on homeassistant/components/camera/arlo.py:
```diff
@@ -158,15 +161,48 @@ def disable_motion_detection(self):
+        base_stations[0].update()
+        self._camera.update()
+
+        battery_level = self.clean_attr(self._camera.get_battery_level)
```
Do you need to drop the `get_` from these based on [this commit](https://github.com/tchellomello/python-arlo/commit/c58d96932c57069d660f5cb9e4d84bc81e7a8ef3)?

mikesplain replied:
Sorry! I'm seeing some weird errors while trying to fix an arlo issue. Thanks!

tchellomello replied:
Yes drop the `get_` from that attribute :+1:  Good point @mikesplain

tchellomello commented on homeassistant/components/sensor/arlo.py:
```diff
@@ -118,6 +132,12 @@ def update(self):
 
+        elif self._sensor_type == 'signal_strength':
+            try:
+                self._state = self._data.get_signal_strength
```
Drop the `get_` from this method and remember to change it too from `battery`

bachya commented:
FYI, since this didn't make it into 0.56.0, opened this bug to track: https://github.com/home-assistant/home-assistant/issues/10016

tchellomello commented on homeassistant/components/camera/arlo.py:
```diff
@@ -158,15 +161,48 @@ def disable_motion_detection(self):
-        self.attrs[
-            ATTR_MOTION] = self._camera.get_motion_detection_sensitivity,
+        # pylint: disable=W0212
+        base_stations = self._camera._session.base_stations
```
@jwillaz  https://pypi.python.org/pypi/pyarlo/0.0.9 has been published.  You can do something like:
```python
base_station = self._camera.base_station
if base_station:
    base_station._refresh_rate = SCAN_INTERVAL.total_seconds()
```

tchellomello commented on homeassistant/components/camera/arlo.py:
```diff
@@ -158,15 +161,48 @@ def disable_motion_detection(self):
+        self._camera.update()
+
+        battery_level = self.clean_attr(self._camera.get_battery_level)
+        brightness = self.clean_attr(self._camera.get_brightness)
```
remove `get_`

tchellomello commented on homeassistant/components/camera/arlo.py:
```diff
@@ -158,15 +161,48 @@ def disable_motion_detection(self):
+
+        battery_level = self.clean_attr(self._camera.get_battery_level)
+        brightness = self.clean_attr(self._camera.get_brightness)
+        flip_state = self.clean_attr(self._camera.get_flip_state)
```
remove `get_`

tchellomello commented on homeassistant/components/camera/arlo.py:
```diff
@@ -158,15 +161,48 @@ def disable_motion_detection(self):
+        flip_state = self.clean_attr(self._camera.get_flip_state)
+        mirror_state = self.clean_attr(self._camera.get_mirror_state)
+        motion_sensitivity = self.clean_attr(
+            self._camera.get_motion_detection_sensitivity)
```
remove `get_`

tchellomello commented on homeassistant/components/camera/arlo.py:
```diff
@@ -158,15 +161,48 @@ def disable_motion_detection(self):
+        mirror_state = self.clean_attr(self._camera.get_mirror_state)
+        motion_sensitivity = self.clean_attr(
+            self._camera.get_motion_detection_sensitivity)
+        powersave_mode = self.clean_attr(self._camera.get_powersave_mode)
```
remove `get_`

tchellomello commented on homeassistant/components/camera/arlo.py:
```diff
@@ -158,15 +161,48 @@ def disable_motion_detection(self):
+        motion_sensitivity = self.clean_attr(
+            self._camera.get_motion_detection_sensitivity)
+        powersave_mode = self.clean_attr(self._camera.get_powersave_mode)
+        signal_strength = self.clean_attr(self._camera.get_signal_strength)
```
remove `get_`

tchellomello commented on homeassistant/components/sensor/arlo.py:
```diff
@@ -97,6 +98,19 @@ def unit_of_measurement(self):
         """Get the latest data and updates the state."""
+        try:
+            # pylint: disable=W0212
+            base_stations = self._data._session.base_stations
```
Use the new `base_station` property

tchellomello commented on homeassistant/components/sensor/arlo.py:
```diff
@@ -97,6 +98,19 @@ def unit_of_measurement(self):
+            return None
+
+        # pylint: disable=W0212
+        base_stations[0]._refresh_rate = SCAN_INTERVAL.total_seconds()
```
same as above

tchellomello commented on requirements_all.txt:
```diff
@@ -568,7 +568,7 @@ pyairvisual==1.0.0
 
 # homeassistant.components.arlo
-pyarlo==0.0.7
+pyarlo==0.0.8
```
to use the new `base_station` attribute, make sure to bump it to `0.0.9`

tchellomello commented on homeassistant/components/camera/arlo.py:
```diff
@@ -158,15 +161,48 @@ def disable_motion_detection(self):
+        # pylint: disable=W0212
+        base_stations[0]._refresh_rate = SCAN_INTERVAL.total_seconds()
+
+        base_stations[0].update()
```
@jwillaz  https://pypi.python.org/pypi/pyarlo/0.0.9 has been published.  You can do something like:
```python
base_station = self._camera.base_station
if not base_station:
    return None

base_station.refresh_rate = SCAN_INTERVAL.total_seconds()
base_station.update()   
```

tchellomello commented on homeassistant/components/camera/arlo.py:
```diff
@@ -158,15 +161,48 @@ def disable_motion_detection(self):
+
+        # pylint: disable=W0212
+        self.attrs[ATTR_LAST_REFRESH] = datetime.fromtimestamp(
+            base_stations[0]._last_refresh).strftime(
```
You can retrieve the `_last_refresh` by doing ```base_station.last_refresh```

jwillaz committed:
Update pyarlo version and review fixes

jwillaz committed:
Fix lint errors

jwillaz committed:
Remove staticmethod

jwillaz commented:
@pvizeli @tchellomello @andrey-git I've made review fixes, updated to the latest pyarlo release (0.1.0), and modified to align with that release. This should be ready to go and will address the exceptions that have been rendering the Arlo platform unusable.

tchellomello commented:
Fixes: https://github.com/home-assistant/home-assistant/pull/9892

tchellomello commented:
Fixes: 10016

lundpetersen commented:
Any status in when this will be implemented?
To me it looks like Arlo implementation is more or less 100% broken at the moment, and @tchellomello and @jwillaz has submitted fixes that will fix the implementation.

bachya commented:
@lundpetersen In the meantime, you can always monkey patch this fix into your current install. For instance, to monkey patch the cameras, copy both `homeassistant/components/arlo.py` and `homeassistant/components/camera/arlo.py` into a `custom_components` directory located within your configuration directory.

lundpetersen commented:
@bachya - I didn't know that - I will try that tonight. Thanks 👍

pvizeli commented on homeassistant/components/camera/arlo.py:
```diff
@@ -113,6 +115,7 @@ def device_state_attributes(self):
             ATTR_POWERSAVE: self.attrs.get(ATTR_POWERSAVE),
             ATTR_SIGNAL_STRENGTH: self.attrs.get(ATTR_SIGNAL_STRENGTH),
             ATTR_UNSEEN_VIDEOS: self.attrs.get(ATTR_UNSEEN_VIDEOS),
+            ATTR_LAST_REFRESH: self.attrs.get(ATTR_LAST_REFRESH),
```
Take it easy without the big part on update:
```
return {
    name: value for name, value in (
       (ATTR_POWERSAVE, POWERSAVE_MODE_MAPPING.get(self._camera.powersave_mode),
       (ATTR_UNSEEN_VIDEOS: self._camera.unseen_videos),
       ...
     ) if value is not None
}
```

pvizeli commented on homeassistant/components/camera/arlo.py:
```diff
@@ -158,15 +161,42 @@ def disable_motion_detection(self):
+        last_refresh = base_station.last_refresh
+        if last_refresh:
+            self.attrs[ATTR_LAST_REFRESH] = datetime.fromtimestamp(
+                last_refresh).strftime("%A, %B %d, %Y %I:%M:%S")
```
How often change this value?

jwillaz replied:
@pvizeli The updates will run every 90 seconds (interval set at the top of the components).  The way these cameras work, however, all of them will fire an update(), and those usually go to the same base station.  So, if you have 4 Arlo / Arlo Pro cameras, every 90 seconds there will be 4 base station update queries, which is a bit unnecessary.  Do you have any ideas for how to reduce that to 1 call from all camera components?  I don't think the Throttle decorator will work, because each component is technically obeying the 90 second interval - they just aren't aware of one another. My assumption is that the only way to fix this concurrency would be to add a locking mechanism in the PyArlo library itself. @tchellomello Thoughts on that?

balloob replied:
We should not store timestamps in Home Assistant state attributes, please remove it.

tchellomello replied:
@jwillaz  I think your idea is good, however use the timmer as an internal attribute instead of an attribute for HASS state.

jwillaz committed:
Clean up attributes

pvizeli committed:
Update arlo.py
